### DISCUSSION TITLE
Application framework architecture

### DISCUSSION THREAD

**Discussion by Rodien** [oqtane/oqtane.framework#2568] [Q&A] [answered]
Hi @sbwalker or anyone with experience,

I have an architecture question. I myself don't have any experience building application frameworks (like DNN or Oqtane). 
But I thought that if there was anyone, I could ask this question, it would be you :) @sbwalker 

Would a design like the one below ever work? 

I was thinking why hasn't anyone built something like this? There has to be a reason?

![image](https://user-images.githubusercontent.com/50739829/213423894-4d726e76-d453-4b04-b4e4-19ea0cd606d1.png)

When working with DNN, taking the production website offline is necessary as the upgrade process involves updating the database. However, separating the application and module information from the core business data can minimize downtime. By keeping the two distinct, new versions of the framework and modules can be tested and swapped without disrupting the live site. This allows for a more efficient and less disruptive upgrade process.

By local database I mean something like SQLite.

However, I must be overlooking something.

**Comment by sbwalker** [comment]
@Rodien I do not understand your diagram... are you suggesting that storing framework data and module data in different databases could reduce production downtime during upgrades?

**Reply by Rodien** [reply]
My thought is that if we store the framework and module information in different databases, we can upgrade the framework and modules in a staging environment, as outlined in https://learn.microsoft.com/en-us/azure/app-service/deploy-staging-slots. This way, we can easily switch between environments without taking the website offline. 

However, currently the framework and module versions and core business data are all stored within the same database, which means that every time we want to upgrade, we have to take the website offline. To avoid this, I was thinking is it not possible to store the framework and module information in one storage and the core business data in another. For example, if we download and install 2sxc, the information would be stored in one database, and any other data, such as news articles, would be inserted into a second database. This way, we can upgrade the application framework and or modules in a staging environment and switch it with the production environment without ever taking it offline.

Again, I have zero experience with building application frameworks, so maybe I am talking nonsense here. 

I was thinking of a way to improve the process of upgrading sites like DNN/ Oqtane and modules. 

But I do not know if it is feasible...

**Comment by sbwalker** [comment]
Sorry for the delayed response... there is indeed a technical reason why this approach does not work in practice. In most applications there is a dependency between application code and data stored in an external database. That dependency is expressed in many different ways depending on the platform, but if we focus on modern .NET then the most common data access approach is EF Core. EF Core relies on models expressed as classes in your application code. So if you change the underlying structure of your data then you need to make the corresponding changes to your code. And if you modify your database then you need to be sure you are running the corresponding application code which is compatible with that database, or else the application will immediately throw errors. And although an external database can be upgraded independently from the application, the application also needs to be upgraded to be compatible - which requires an application restart. This is required because the app domain needs to be unloaded and reloaded with the new assemblies. And this is because .NET will only register your services for dependency injection during app startup.

**Comment by sbwalker** [comment, marked as answer]
To clarify, I am not saying that it is impossible to avoid downtime with Oqtane... it is definitely possible to achieve this. However it is not a framework responsibility - it is an infrastructure responsibility. The Blue/Green concept has been around for more than a decade for exactly this scenario: https://martinfowler.com/bliki/BlueGreenDeployment.html

**Reply by Rodien** [reply]
Okay thank you @sbwalker :)

I will need to take a look at the following: 

"Databases can often be a challenge with this technique, particularly when you need to change the schema to support a new version of the software. The trick is to separate the deployment of schema changes from application upgrades. So first apply a database refactoring to change the schema to support both the new and old version of the application, deploy that, check everything is working fine so you have a rollback point, then deploy the new version of the application. (And when the upgrade has bedded down remove the database support for the old version.)" - Martin fowler 

Fowler recommended the book 'Refactoring Databases'
